Worked on the RFXCom module

martinvw opened the pull request:
~~~Add~~~ *Fixed* supported for Chime and genericly updated the mapping from bytes to ENUM's
Added a test package with coverage for some of the messages
Solved some array size / message length related problems *(which were found by added tests)*

See also: https://github.com/openhab/openhab/pull/4629

martinvw commented:
I decided to exclude the renaming of the SubTypes but forgot the test

martinvw commented:
He guys, anything I could or should do to make sure the pull requests can be reviewed and merged?

Jamstah commented:
I'd recommend breaking this out into more focused commits. Maybe one for the tests, one for adding new support, one for fixing bugs, one for swapping to enums, whatever makes sense such that each individual commit (and pull request) stands on its own as an improvement.

As it is, there is a lot to review in one go, and for me, it makes it hard to understand the reasons behind some of the changes.

martinvw commented:
Thanks for your response, splitting the pull request is possible but requires things to be merged (quicker). I could split up in the following steps:

1. Add skeleton for tests (including tests that succeed on master branch)
1. Add additional tests including the code changes that make them succeed (requires the first step to be merged, can be split up in small pull requests, will be approx 6 pull requests)
1. ...
1. Use the proper byte to enum mapping everywhere (requires merging of above mentioned pull requests) this makes it likely that this same set of errors will not be re-introduced.

Is this something that could work?

And what can I do to make sure that the (not yet created first pull request) be merged quick because otherwise I cannot proceed in a sensible way to the next steps?

paulianttila commented:
I quickly checked the changes and for me they look good. Great to have unit tests for rfxcom binding.

Jamstah commented on addons/binding/org.openhab.binding.rfxcom.test/.project:
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>org.openhab.binding.max.test</name>
```
Wrong project name

Jamstah commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComBlinds1Message.java:
```diff
@@ -1,6 +1,6 @@
 /**
  * Copyright (c) 2014-2016 by the respective copyright holders.
- *
+ * <p>
```
Does this need to be here?

Jamstah commented on addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComTemperatureHumidityMessage.java:
```diff
@@ -90,7 +109,7 @@ public byte toByte() {
     private final static List<RFXComValueSelector> supportedOutputValueSelectors = Arrays.asList();
 
-    public SubType subType = SubType.UNDEF;
+    public SubType subType = SubType.TH1;
```
Why are we selecting a default subtype instead of starting with unknown? Until the message has been encoded, this value could be incorrect (applies to a couple of places).

paulianttila replied:
Subtype UNDEF (0) doesn't follow the RFXCOM spec (it's fictitious subtype).

Jamstah replied:
I agree with removing the undef, especially if it's not in spec, but they all also have UNKNOWN(255). I thought initializing with that might be more accurate.

paulianttila replied:
I agree, that's why I introduced UNKNOWNS, even they not in the rfxcom specs either.

martinvw replied:
I will convert all initialization to UNKNOW

Jamstah commented:
As a personal opinion, I'd have said put all the tests in one commit with all the changes required to fix any bugs discovered by the new tests. And I fully agree with Pauli, great to have unit tests at all. I'd have said the new content (chimes) and enum work deserve their own commits because they are logically separate.

This is personal opinion on "neat" development though, Kai / Pauli may have different opinions. If you are keeping it as one block of work, you should squash the two commits in this pull request and make sure you keep the Signed Off By line in the comment. If you split it up, and the commits depend on each other (but have logically separate objectives), you can probably just have all the commits in one pull request for practicalities sake. I tend to use branching to try and avoid overlapping work myself, although merging is an unwelcome side effect there :smile:

martinvw commented:
The base of unit tests (especially the classes RFXComTemperatureHumidityMessageTest & RFXComMessagesTest) are ported from openhab 1 and are created by @ivanfmartinez

@paulianttila thanks for your time. Anyone else who should take a look at it for it to be accepted?

martinvw commented:
@Jamstah Thanks for your time, I think I should be able to fix the issues tonight.

Btw. The chime support is not new but was broken so its also the result of fixing the code because of a failed test. To fix the generic mismatch problem I changed the way byte's are mapped to the enum values on all locations.

Jamstah commented:
I must have misread your commit comment then. If it's all fixes related to the tests then one commit is great :smile:

martinvw commented:
I updated the initial pull request message to make it more clear and fixed the review comments. If the build is green and there are no further remarks I can squash the commits and it should be ready for merging.

Jamstah commented:
Lgtm. @kaikreuzer     .

martinvw commented:
@paulianttila, @Jamstah thanks for your efforts. There is a more or less similar pull-request for OH1. If you have any time in the upcoming weeks it would be greatly appreciated.

martinvw commented:
The openHAB 1 pull requests has been merged successfully.

martinvw commented:
Note to self, convertSubType still uses the old way to retrieve subTypes, this code is OH2 specific so I missed it. 

@kaikreuzer missing this small change does not make it disfunctional, only some bugs might not yet be resolved. So I can also resolve it in a separate pull request, which makes them easier to review. Do you have any opinion about this?

kaikreuzer commented on addons/binding/org.openhab.binding.rfxcom.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,19 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Openhab RFXCOM Binding Tests
```
openHAB

kaikreuzer commented on addons/binding/org.openhab.binding.rfxcom.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,19 @@
+Bundle-SymbolicName: org.openhab.binding.rfxcom.test;singleton:=true
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-Vendor: Openhab
```
openHAB

kaikreuzer commented:
Ah, so much nicer - thanks :-)

martinvw commented:
And done, sorry for mis-spelling :-D
